### PR TITLE
Add a part of unit tests for `builder_state.rs` to legacy builder

### PR DESCRIPTION
**Pull request by dailinsubjam**
Closes some of https://github.com/EspressoSystems/marketplace-builder-core/issues/133 .

### This PR:
- Adds tests for `process_da_proposal`, `process_quorum_proposal` and `process_decide_event` for `builder_state.rs` in legacy builder.
- Adds testing framework for legacy builder.

### This PR does not:
- Add test for `spawn_clone`, `build_block` and `process_block_request` yet.

### Key places to review:
All changes.

### How to test this PR: 
`RUSTFLAGS="--cfg async_executor_impl=\"tokio\"" RUST_LOG=info cargo test  -- --nocapture`

### CODE REVIEW

File: crates/legacy/src/builder_state.rs [hunk 1]
```diff
@@ -1143,3 +1141,414 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
...
+                    leaf_commit,
+                    builder_commitment: quorum_proposal.block_header.builder_commitment,
+                },
+                req_sender,
+            );
+        }
+
+        // send out a decide event
+        // randomly choose a latest_decide_view_number between [0, NUM_ROUNDS]
+        let latest_decide_view_number = ViewNumber::new(3);
```
Comment (QuentinI) [inline crates/legacy/src/builder_state.rs:1533]: image

I think the comment is a bit misleading here 😄

File: crates/legacy/src/builder_state.rs [hunk 2]
```diff
@@ -1143,3 +1141,414 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
...
+        // Capacity of broadcast channels
+        const CHANNEL_CAPACITY: usize = NUM_ROUNDS * 5;
+        // Number of nodes on DA committee
+        const NUM_STORAGE_NODES: usize = 4;
+
+        // start builder_state without entering event loop
+        let (_senders, global_state, mut builder_state) =
+            start_builder_state_without_event_loop(CHANNEL_CAPACITY, NUM_STORAGE_NODES).await;
+
+        // insert some builder states
```
Comment (QuentinI) [inline crates/legacy/src/builder_state.rs:1490]: This comment feels unrelated to any of the surrounded code
Is there anything it applies to or is it here by mistake?

File: crates/legacy/src/legacy_testing/mod.rs [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+/// get transactions submitted in previous rounds, [] for genesis
+/// and simulate the block built from those
+pub async fn calc_proposal_msg(
+    num_storage_nodes: usize,
+    round: usize,
+    prev_quorum_proposal: Option<QuorumProposal<TestTypes>>,
+    transactions: Vec<TestTransaction>,
+) -> (
+    QuorumProposal<TestTypes>,
+    MessageType<TestTypes>,
```
Comment (QuentinI) [inline crates/legacy/src/legacy_testing/mod.rs:129]: Nit: can we replace these two `MessageType`s with underlying concrete message types and wrap them at the call site? I think this would make function signature more clear at a glance

File: crates/legacy/src/legacy_testing/mod.rs [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+            &block_payload,
+            &metadata,
+        );
+
+    // generate key for leader of this round
+    let seed = [round as u8; 32];
+    let (pub_key, private_key) = BLSPubKey::generated_from_seed_indexed(seed, round as u64);
+
+    // Prepare the DA proposal message
+    let da_proposal_message: DaProposalMessage<TestTypes> = {
```
Comment (QuentinI) [inline crates/legacy/src/legacy_testing/mod.rs:151]: Is this separate scope needed? I see that it's shadowing `seed` and keys, but it looks like it shadows them with exact same values

**Review by QuentinI — CHANGES_REQUESTED** [review]

File: crates/legacy/src/lib.rs [hunk 5]
```diff
@@ -19,7 +19,7 @@ pub mod service;
 
 // tracking the testing
 #[cfg(test)]
-pub mod testing;
+pub mod legacy_testing;
```
[the remark this answers is not in the document]
Comment (dailinsubjam) [reply crates/legacy/src/lib.rs]: solved in [771cc43498258dae5a4e9c6bd837530204c98d95](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/771cc43498258dae5a4e9c6bd837530204c98d95)

File: crates/legacy/src/builder_state.rs [hunk 6]
```diff
@@ -1143,3 +1141,414 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
...
+    use super::MessageType;
+    use super::ParentBlockReferences;
+    use crate::legacy_testing::{
+        calc_builder_commitment, calc_proposal_msg, start_builder_state_without_event_loop,
+        TestTypes,
+    };
+
+    /// check whether the `da_proposal_payload_commit_to_da_proposal` has correct (key, value) pair after processing da proposal messages
+    /// used for testing only
+    fn check_equal_da_proposal_hashmap<TYPES: NodeType>(
```
[the remark this answers is not in the document]
Comment (dailinsubjam) [reply crates/legacy/src/builder_state.rs:1169]: Weird, previously there will be some type mismatching error. Therefore I wrote a function, but now `assert_eq!()` could work. Fixed in [57cdba0](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/57cdba0caf7a5ef7f7b675c023f23e8b7445de10).

File: crates/legacy/src/legacy_testing/mod.rs [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+use committable::{Commitment, CommitmentBoundsArkless, Committable};
+use marketplace_builder_shared::block::{BuilderStateId, ParentBlockReferences};
+use std::sync::Arc;
+use std::time::Duration;
+
+#[derive(
+    Copy, Clone, Debug, Default, Hash, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
+)]
+pub struct TestTypes;
+impl NodeType for TestTypes {
```
[the remark this answers is not in the document]
Comment (dailinsubjam) [reply crates/legacy/src/legacy_testing/mod.rs]: Previously I just copy-pasted from marketplace `mod.rs`, but it's true that we can remove it. Fixed in [fadf5ee](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/fadf5ee27934afb6666635875f23be417bdcd3a0).

File: crates/legacy/src/legacy_testing/mod.rs [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    type SignatureKey = BLSPubKey;
+    type Transaction = TestTransaction;
+    type ValidatedState = TestValidatedState;
+    type InstanceState = TestInstanceState;
+    type Membership = StaticCommittee<Self>;
+    type BuilderSignatureKey = BuilderKey;
+    type AuctionResult = TestAuctionResult;
+}
+
+pub async fn start_builder_state_without_event_loop(
```
[the remark this answers is not in the document]
Comment (dailinsubjam) [reply crates/legacy/src/legacy_testing/mod.rs]: Renamed to `create_builder_state` in [3558ff8](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/3558ff8f6c99675708a973bfcfd6e7c64f863090)

File: crates/legacy/src/builder_state.rs [hunk 9]
```diff
@@ -1143,3 +1141,414 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
...
+        } else {
+            panic!("global_state should have cooresponding builder_state_id as now we have matching da proposal.");
+        }
+    }
+
+    /// This test the function `process_decide_event`.
+    /// It checkes whether we exit out correct builder states when there's a decide event coming in.
+    /// This test also checks whether corresponding BuilderStateId is removed in global_state.
+    #[async_std::test]
+    async fn test_legacy_process_decide_event() {
```
[the remark this answers is not in the document]
Comment (dailinsubjam) [reply crates/legacy/src/builder_state.rs]: Oops this is just for having different name with the test in marketplace crate for debugging. Changed the name in [41dccb1](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/41dccb1a7b403c26035a0282b0bca5a45e27280d)

File: crates/legacy/src/builder_state.rs [hunk 10]
```diff
@@ -1143,3 +1141,414 @@ impl<TYPES: NodeType> BuilderState<TYPES> {
...
+            panic!("Not a da_proposal_message in correct format");
+        }
+        check_equal_da_proposal_hashmap(
+            builder_state
+                .da_proposal_payload_commit_to_da_proposal
+                .clone(),
+            correct_da_proposal_payload_commit_to_da_proposal.clone(),
+        );
+        // check global_state didn't change
+        if let Some(_x) = global_state
```
[the remark this answers is not in the document]
Comment (dailinsubjam) [reply crates/legacy/src/builder_state.rs]: Updated in [ba55c2a](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/ba55c2aad656796d7ccf8d56298c1fe767a07739)

File: crates/legacy/src/builder_state.rs (hunk 1 above)
Comment (dailinsubjam) [reply crates/legacy/src/builder_state.rs:1478]: [18befc0](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/18befc025d8ee08aa023f33e4301b464366095b8)

File: crates/legacy/src/builder_state.rs (hunk 2 above)
Comment (dailinsubjam) [reply crates/legacy/src/builder_state.rs]: Changed to "// register some builder states for later decide event". This comment is about the following whole block, but I moved it to the place after transaction generation to avoid confusion. [1d14f1f](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/1d14f1f81eacfc86ec8baf62394d5ca28161c825)

File: crates/marketplace/src/builder_state.rs
Comment (shenkeyao) [inline crates/marketplace/src/builder_state.rs:1]: The comment updates make tests clearer. 👍

**Review by shenkeyao — APPROVED** [review]
LGTM. It'd be nice to have @QuentinI take another look in case any of his previous comments isn't fully resolved.

File: crates/legacy/src/legacy_testing/mod.rs (hunk 3 above)
Comment (dailinsubjam) [reply crates/legacy/src/legacy_testing/mod.rs]: Updated in [1d14f1f](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/1d14f1f81eacfc86ec8baf62394d5ca28161c825)

File: crates/legacy/src/legacy_testing/mod.rs (hunk 4 above)
Comment (dailinsubjam) [reply crates/legacy/src/legacy_testing/mod.rs]: Good catch! Didn't notice that when I put them together. Removed in [75165e3](https://github.com/EspressoSystems/marketplace-builder-core/pull/176/commits/75165e35aa792efb87671ef150cb5af00ddc87cd)

**Review by QuentinI — APPROVED** [review]
LGTM! Thanks for links to commits when resolving comments, makes reviewing much easier
